MarianInterface.cpp

```diff
@@ -28,58 +30,57 @@ marian::Ptr<marian::Options> MakeOptions(QString path_to_model_dir, translateLoc
-        using marian::bergamot::Response;
-        QString translation;
-
-        // If translating empty string, just pass through immediately but do
-        // update the finished_ variable to cancel out all previous translations.
-        if (input.empty()) {
-            translation = "";
-        } else {
-            std::future<marian::bergamot::Response> responseFuture = service_->translate(std::move(input));
+    worker_ = std::thread([&](marian::Ptr<marian::Options> options) {
+        marian::bergamot::Service service(options);
+
+        while (true) {
+            // Wait for work
+            pendingInputCount_.acquire();
+
+            std::unique_ptr<std::string> input(pendingInput_.fetchAndStoreAcquire(nullptr));
+            // Empty ptr? pendingInputCount released without valid pointer -> poison (or a bug)
+            if (!input)
+                break;
+
+            emit pendingChanged(true);
+
+            std::future<marian::bergamot::Response> responseFuture = service.translate(std::move(*input));
             responseFuture.wait();
             marian::bergamot::Response response = responseFuture.get();
-            translation = QString::fromStdString(response.target.text);
-        }
+            emit translationReady(QString::fromStdString(response.target.text));
 
-        // There is no guarantee that we get/process responses in the same order
-        // as we sent sentences to be translated. So let's make sure we haven't
-        // been overtaken by a further progressed sentence before emitting the
-        // translation.
-        if (serial < finished_)
-            return;
-        
-        finished_ = serial;
+            emit pendingChanged(false);
+        }
 
-        emit translationReady(translation);
+        // We need to manually destroy the loggers, as marian doesn't do that.
+        spdlog::drop("general");
+        spdlog::drop("valid");
+    }, MakeOptions(path_to_model_dir, settings));
+}
 
-        if (serial_ == finished_)
-            emit pendingChanged(false);
-    };
+void MarianInterface::translate(QString in) {
+    std::unique_ptr<std::string> old(pendingInput_.fetchAndStoreAcquire(new std::string(in.toStdString())));
 
-    std::size_t serial = ++serial_;
-    emit pendingChanged(true);
-    std::thread mythread(translateAndSignal, in.toStdString(), serial);
-    mythread.detach();
+    // notify worker (but only if there wasn't already a pending task)
+    if (!old)
+        pendingInputCount_.release();
 }
 
 MarianInterface::~MarianInterface() {
-    // We need to manually destroy the loggers, as marian doesn't do that.
-    spdlog::drop("general");
-    spdlog::drop("valid");
+    // Poision the worker with a nullptr
+    std::unique_ptr<std::string> old(pendingInput_.fetchAndStoreAcquire(nullptr));
+    if (!old)
+        pendingInputCount_.release();
+    
+    // Wait for worker to join as it depends on resources we still own.
```

Comment:
This is not an issue in the current implementation as we do a thread::detach, but potentially could lead to crashes. I think we disable (or at least I used to) disable changing the model while translation is still ongoing...

Reply:
I'm tempted to add a "detachAndFree()" method or something to MarianInterface. That then basically starts a separate thread that does nothing but `delete this`. When changing model, we could do something like `translator_.release()->detachAndFree()` and just let the MarianInterface object take care of cleaning up itself in a non-main thread-blocking way.

Reply:
That means the worker thread must somehow know it's not "relevant" so that it doesn't emit any signals that mess up the GUI.

Reply:
Yepp, this is the spdlog issue that i have in the destructor.